### PR TITLE
TranslationEngine: fix possible segfault

### PR DESCRIPTION
**Pull request by Lord-Kamina**
@ooshlablu reported seeing a crash on macOS after my second Unicode PR. I haven’t been able to reproduce it but from the trace, I’m thinking the issue must be that it’s trying to use “m_currentLanguage” before the class has finished constructing.

If so, it can probably be fixed by giving it a default value.

### More

- [ ] Added/updated documentation

### CODE REVIEW

**Review by twollgam — APPROVED** [review]
If this solution works merge it. LGTM

**Comment by Lord-Kamina** [comment]
Turns out I was wrong. The real problem was `getCurrentLanguageCode`was being called form the `MidiParser`, which is constructed and put to work before instantiating fake singleton, resulting dereferencing an empty pointer. I made TranslationEngine static, removing its reliance on `Game` altogether.

@ooshlablu can you confirm if it works for you?

**Comment by ooshlablu** [comment]
This fixes the segfault I was seeing on MacOS

**Review by ooshlablu — APPROVED** [review]
LGTM

File: game/songparser.cc [hunk 1]
```diff
@@ -70,16 +70,17 @@ SongParser::SongParser(Song& s): m_song(s) {
 			type = Type::XML;	// XMLPP should deal with encoding so we don't have to.
 		}
 		else {
-			UnicodeUtil::convertToUTF8(m_ss.str(), s.filename.string());
-			if (smCheck (m_ss.str())) {type = Type::SM; } else if (txtCheck (m_ss.str())) {
+			std::string ss = UnicodeUtil::convertToUTF8(m_ss.str(), s.filename.string());
+			if (smCheck (ss)) {type = Type::SM; } else if (txtCheck (ss)) {
```
Comment (twollgam) [inline game/songparser.cc:74]: I would like to see some line breaks ;-)

File: game/unicode.cc [hunk 2]
```diff
@@ -36,10 +36,11 @@ Converter& UnicodeUtil::getConverter(std::string const& s) {
 	return m_converters.try_emplace(s, Converter(s)).first->second;
 }
 
-std::string UnicodeUtil::getCharset (std::string_view str) {
+std::string UnicodeUtil::getCharset (std::string_view& str) {
```
Comment (twollgam) [inline game/unicode.cc:39]: Why (non const) reference here? Is str modified?

File: game/unicode.cc [hunk 3]
```diff
@@ -98,7 +96,7 @@ std::string UnicodeUtil::convertToUTF8 (std::string_view str, std::string _filen
 	return ret;
 }
 
-bool UnicodeUtil::removeUTF8BOM(std::string_view str) {
+bool UnicodeUtil::removeUTF8BOM(std::string_view& str) {
```
Comment (twollgam) [inline game/unicode.cc:99]: Reference needed?

File: game/songparser.cc (hunk 1 above)
Comment (Lord-Kamina) [reply game/songparser.cc:74]: Not saying I won't add them but man I seriously want to know what's wrong with your editor width setting 😂

File: game/unicode.cc (hunk 2 above)
Comment (Lord-Kamina) [reply game/unicode.cc:39]: It can be modified by removeUTF8BOM

**Review by twollgam — APPROVED** [review]

File: game/unicode.cc (hunk 3 above)
Comment (Lord-Kamina) [reply game/unicode.cc:99]: Yes because it potentially modifies the string_view to a substring of itself.

File: game/i18n.cc [hunk 4]
```diff
@@ -120,19 +125,24 @@ std::string TranslationEngine::getLanguageByKey(const std::string& languageKey)
...
 
-const std::pair<std::string, std::string>& TranslationEngine::getCurrentLanguage() const { 
-	return m_currentLanguage; 
+std::pair<std::string, std::string> const& TranslationEngine::getCurrentLanguage() { 
+	return TranslationEngine::m_currentLanguage; 
+}
+
+std::string TranslationEngine::getCurrentLanguageCode() {
+	std::string lang(getCurrentLanguage().first);
+	return lang.substr(0, lang.size() - 6);
```
Comment (Baklap4) [inline game/i18n.cc:134]: Whats this minus 6??

**Review by Baklap4 — COMMENTED** [review]
Overall looks alright
Is the change to static functions really needed? Im always getting itchy if a lot of static is used...

**Comment by Lord-Kamina** [comment]
There are other possible solutions, but.I think they'd require a deeper refactoring and/or to keep relying on`Game`

File: game/i18n.cc (hunk 4 above)
Comment (Lord-Kamina) [reply game/i18n.cc:134]: Sorry for the magic number, it's ".UTF-8"

Comment (Lord-Kamina) [reply game/i18n.cc:134]: Maybe it'd be best to calculate the cut-off in relation to the index of the first `.`?

**Review by Baklap4 — APPROVED** [review]
LGTM, still on holiday, but people claimed it fixes the issue :)

**Pull request merged**